# Домашнее задание к лекции «Случайные события. Случайные величины»

Смоделировать игру против лотерейного автомата типа “777”. Игрок платит 1 руб., после чего выпадает случайное целое число, равномерно распределенное от 0 до 999. При некоторых значениях числа игрок получает выигрыш (см. справа)

Выгодна ли игра игроку?

Сколько в среднем приобретает
или теряет игрок за одну игру?

777: 200 руб.

999: 100 руб.

555: 50 руб.

333: 15 руб.

111: 10 руб.

*77: 5 руб.

**7: 3 руб.

*00: 2 руб.

**0: 1 руб.

– любая цифра
Дополнительное задание повышенной сложности. Теоретически рассчитать средний выигрыш (проигрыш) и сравнить с результатами моделирования

In [2]:
import numpy as np


In [2]:
np.random.seed(777)
trials = 1000000 # количество игр
wins = {} # количество выигрышей по категориям
bet = 1 # ставка

def win_count(category):
    '''Подсчет количества выигышей по категориям и запись в словарь wins'''
    category = tuple(category)
    if category in wins.keys():
        wins[category] += 1
    else:
        wins[category] = 1



for i in range(trials):
    slots = [np.random.randint(0, 10) for i in range(3)]
    
    if slots[0]==slots[1]==slots[2]==7:
        win_count(('777',200))
        
    elif slots[0]==slots[1]==slots[2]==9:
        win_count(('999', 100))
        
    elif slots[0]==slots[1]==slots[2]==5:
        win_count(('555', 50))
        
    elif slots[0]==slots[1]==slots[2]==3:
        win_count(('333', 15))
        
    elif slots[0]==slots[1]==slots[2]==1:
        win_count(('111', 10))
        
    elif slots[0]!=7 and slots[1]==slots[2]==7:
        win_count(('*77', 5))
        
    elif slots[0]!=7 and slots[1]!=7 and slots[2]==7:
        win_count(('**7', 3))

    elif slots[1]==slots[2]==0:
        win_count(('*00', 2))

    elif slots[2]==0:
        win_count(('**0', 1))
    
all_bets = bet * trials

all_wins = 0
for k,v in wins.items():
    win =(k[1]*v)
    all_wins += win

# общий выигрыш за минусом ставок
all_wins = all_wins - all_bets
print(f'trials = {trials}    all_bets = {all_bets}    all_win = {all_wins}')
l=len(f'trials = {trials}    all_bets = {all_bets}    all_win = {all_wins}')
print('_'*l)
# определение полной вероятности выйгрыша по методу Монте-Карло
# посчитать полную вероятность выйгрыша P(A)=P(H1)+P(H2)+...
# вероятность выигрыша по категории P(H) = win_count(H)/trials
# 

P_A = 0
win_aver = 0

for k, v in wins.items():
    P = v/trials
    P_A += P
    win_aver += P * k[1]
    print(f'Вероятность P\({k[0]}\)={P}')
print('_'*l)
print(f'Вероятность выигрыша P(A) = {P_A}')
print()
print(f'Средний выигрыш за минусом ставки = {win_aver - bet}')


trials = 1000000    all_bets = 1000000    all_win = -229947
___________________________________________________________
Вероятность P\(**0\)=0.089769
Вероятность P\(**7\)=0.0807
Вероятность P\(555\)=0.000988
Вероятность P\(*00\)=0.010127
Вероятность P\(*77\)=0.009059
Вероятность P\(111\)=0.001049
Вероятность P\(777\)=0.000992
Вероятность P\(333\)=0.000983
Вероятность P\(999\)=0.000996
___________________________________________________________
Вероятность выигрыша P(A) = 0.19466299999999997

Средний выигрыш за минусом ставки = -0.229947


## Вывод:

Игра невыгодна игроку.

Средний проигрыш за игру составляет 0.229947 руб.

In [10]:
np.random.seed(777)
trials = 1000 # количество игр
wins = {} # количество выигрышей по категориям
bet = 1 # ставка

def win_count(category):
    '''Подсчет количества выигышей по категориям и запись в словарь wins'''
    category = tuple(category)
    if category in wins.keys():
        wins[category] += 1
    else:
        wins[category] = 1



for i in range(trials):
    slots = [np.random.randint(0, 10) for i in range(3)]
    
    if slots[0]==slots[1]==slots[2]==7:
        win_count(('777',200))
        
    elif slots[0]==slots[1]==slots[2]==9:
        win_count(('999', 100))
        
    elif slots[0]==slots[1]==slots[2]==5:
        win_count(('555', 50))
        
    elif slots[0]==slots[1]==slots[2]==3:
        win_count(('333', 15))
        
    elif slots[0]==slots[1]==slots[2]==1:
        win_count(('111', 10))
        
    elif slots[0]!=7 and slots[1]==slots[2]==7:
        win_count(('*77', 5))
        
    elif slots[0]!=7 and slots[1]!=7 and slots[2]==7:
        win_count(('**7', 3))

    elif slots[1]==slots[2]==0:
        win_count(('*00', 2))

    elif slots[2]==0:
        win_count(('**0', 1))
    
all_bets = bet * trials
all_wins = sum([(k[1]*v) for k,v in wins.items()])

# общий выигрыш за минусом ставок
all_wins = all_wins - all_bets
print(f'trials = {trials}    all_bets = {all_bets}    all_win = {all_wins}')
l=len(f'trials = {trials}    all_bets = {all_bets}    all_win = {all_wins}')
print('_'*l)
# определение полной вероятности выйгрыша по методу Монте-Карло
# посчитать полную вероятность выйгрыша P(A)=P(H1)+P(H2)+...
# вероятность выигрыша по категории P(H) = win_count(H)/trials
# средний выигрыш win = P1X1 + P2X2 + ...

P_A = 0
win_aver = 0

for k, v in wins.items():
    P = v/trials
    P_A += P
    win_aver += P * k[1]
    print(f'Вероятность P({k[0]})={P}')

print('_'*l)
print(f'Вероятность выигрыша P(A) = {P_A}')
print()
print(f'Средний выигрыш за минусом ставки = {win_aver - bet}')


trials = 1000    all_bets = 1000    all_win = -264
__________________________________________________
Вероятность P(**0)=0.092
Вероятность P(**7)=0.091
Вероятность P(555)=0.002
Вероятность P(*00)=0.013
Вероятность P(*77)=0.005
Вероятность P(111)=0.002
Вероятность P(777)=0.001
__________________________________________________
Вероятность выигрыша P(A) = 0.20600000000000002

Средний выигрыш за минусом ставки = -0.264
